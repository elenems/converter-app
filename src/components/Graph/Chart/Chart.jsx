import React, { Component } from "react";
import ChartTable from "chart.js";
import { getCurrencyData } from "../utils";

let chart = null;
let ctx = null;

export default class Chart extends Component {
  componentDidMount() {
    ctx = document.getElementById("myChart").getContext("2d");
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.props.historyRates !== nextProps.historyRates) {
      return true;
    }
    return false;
  }

  componentDidUpdate() {
    if (chart !== null) {
      chart.destroy();
    }

    const { historyRates, currencyFrom, currencyTo } = this.props;
    const { differences, dates } = getCurrencyData(
      historyRates,
      currencyFrom,
      currencyTo
    );

    chart = new ChartTable(ctx, {
      type: "line",
      data: {
        labels: dates,
        datasets: [
          {
            label: "Currency difference",
            data: differences,
            pointBorderColor: "rgba(0,0,0,0)",
            pointBackgroundColor: "rgba(0,0,0,0)",
            backgroundColor: ["rgba(0,0,0,0)"],
            borderColor: ["#32a1ce"],
            borderWidth: 1
          }
        ]
      },
      options: {
        maintainAspectRatio: false,
        scales: {
          yAxes: [
            {
              ticks: {
                beginAtZero: true
              }
            }
          ],
          xAxes: [
            {
              type: "time",
              time: {
                unit: "month"
              }
            }
          ]
        }
      }
    });
  }

  render() {
    return <canvas id="myChart"></canvas>;
  }
}
