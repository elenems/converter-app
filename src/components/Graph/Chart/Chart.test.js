import React from "react";
import { shallow, mount } from "enzyme";
import Chart from "./Chart";

describe("Chart", () => {
  beforeAll(() => {
    const canvas = document.createElement('canvas');
    canvas.setAttribute('id', 'myChart');
    document.body.appendChild(canvas);
  });
  
  afterAll(() => {
    const canvas = document.getElementById('myChart');
    if (canvas) {
      document.body.removeChild(canvas);
    }
  });

  it('should render correctly in "debug" mode', () => {
    const component = shallow(<Chart />).debug();
    expect(component).toMatchSnapshot();
  });

  it("mounts component", () => {
    const wrapper = shallow(<Chart />);
    const call = jest.spyOn(wrapper.instance(), "componentDidMount");
    wrapper.instance().componentDidMount();
    expect(call).toHaveBeenCalled();
  });

  it("updates component", () => {
    const wrapper = shallow(<Chart historyRates={{}} />);
    const nextProps = {};
    const call = jest.spyOn(wrapper.instance(), "shouldComponentUpdate");
    wrapper.instance().shouldComponentUpdate(nextProps);
    expect(call).toHaveBeenCalled();
  });

  it("not update on same props", () => {
    const props = {
      historyRates: {}
    };
    const wrapper = shallow(<Chart {...props} />);
    const call = jest.spyOn(wrapper.instance(), "shouldComponentUpdate");
    const result = wrapper.instance().shouldComponentUpdate(props);
    expect(call).toHaveReturnedWith(false);
  });

  it("components updates", () => {
    const props = { historyRates: [] };
    const nextProps = {historyRates: [1]}
    const wrapper = mount(<Chart {...props} />, {attachTo: document.getElementById('myChart')});
    const fn = jest.spyOn(wrapper.instance(),'componentDidUpdate');
    wrapper.setProps(nextProps);
    expect(fn).toHaveBeenCalled()
  });
});
