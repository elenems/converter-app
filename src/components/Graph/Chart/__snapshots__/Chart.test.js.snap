// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;
